src/Nethermind/Nethermind.Synchronization/Peers/PeerInfo.cs

```diff
@@ -183,31 +184,38 @@ private static string BuildContextString(AllocationContexts contexts)
         {
             return $"{((contexts & AllocationContexts.Headers) == AllocationContexts.Headers ? "H" : " ")}{((contexts & AllocationContexts.Bodies) == AllocationContexts.Bodies ? "B" : " ")}{((contexts & AllocationContexts.Receipts) == AllocationContexts.Receipts ? "R" : " ")}{((contexts & AllocationContexts.State) == AllocationContexts.State ? "S" : " ")}";
         }
-        
-        public override string ToString() => $"[{BuildContextString(AllocatedContexts)}][{BuildContextString(SleepingContexts)}]{SyncPeer}";
 
+        public override string ToString()
+        {
+            return $"[{BuildContextString(AllocatedContexts)}][{BuildContextString(SleepingContexts)}]{SyncPeer}";
+        }
+
+        /// <summary>
+        /// Determines the type of client on behalf of provided clientId (aka UserAgent)
+        /// </summary>
+        /// <remarks>
+        ///  Mantain labels of <see cref="PeerClientType"/> enum object accordingly.
+        /// </remarks>
+        /// <param name="syncPeer">The peer being recognized</param>
         private void RecognizeClientType(ISyncPeer syncPeer)
         {
-            if (syncPeer.ClientId?.Contains("BeSu", StringComparison.InvariantCultureIgnoreCase) ?? false)
-            {
-                PeerClientType = PeerClientType.BeSu;
-            }
-            else if (syncPeer.ClientId?.Contains("Geth", StringComparison.InvariantCultureIgnoreCase) ?? false)
-            {
-                PeerClientType = PeerClientType.Geth;
-            }
-            else if (syncPeer.ClientId?.Contains("Nethermind", StringComparison.InvariantCultureIgnoreCase) ?? false)
-            {
-                PeerClientType = PeerClientType.Nethermind;
-            }
-            else if (syncPeer.ClientId?.Contains("Parity", StringComparison.InvariantCultureIgnoreCase) ?? false)
+            if (!string.IsNullOrEmpty(syncPeer.ClientId))
             {
-                PeerClientType = PeerClientType.Parity;
-            }
-            else
-            {
-                PeerClientType = PeerClientType.Unknown;
+                // Assume Unknown is value 255
+                Dictionary<string, int> clientTypes = Enum.GetValues(typeof(PeerClientType)).Cast<PeerClientType>().Where(t => (int)t != 255).ToDictionary(k => k.ToString(), v => (int)v);
```

Comment:
This line allocates a lot of memory by creating dictionary and Where iterators
The code above is not bad coding - it is optimized to avoid allocations

Enum.GetValues will allocatem Where will allocate itertor. Magic value of 255 should not be used.
To DIctionary allocates. Calling ToString allocates a string. Looks like ~600bytes of allocations

Reply:
The number of times this is called is very low. The trade-off is to decide whether the code should be amended every time the enum changes ( adding if branches) or simply add elements to enum and let enum object be reflected.

Reply:
Eventually dictionary can be set as static

Reply:
I think it may be called a few hundred times per second when peering which would put some GC pressure. Can use a Dictionary from outside of method but probably not worth. So the options are that the PR is perfect and aproved or we just leave it as it is and the code would be improved the next time we add a new client to the list.